### ISSUE TITLE
Enforce that IR crossing the backend plugin API is only `stablehlo`.

### ISSUE THREAD

**Issue by hawkinsp** [google/jax#15608]
Currently JAX emits a mixture of `chlo` and `stablehlo` and expects the backend (plugin) to handle it.

`chlo` does not promise any form of backward or forward compatibility, so to have a stable plugin API JAX needs to lower away any `chlo` pieces before handing them to the backend.

**Comment by jpienaar** [comment]
The mix is actually a preferred situation for OpenXLA today as CHLO has some better dynamic shape support (as well as composed ops that have different libraries/expansions for different backbends that would have to be matched back up if lowered to StableHLO), vice versa StableHLO has some composed ops that probably should be in CHLO.

We'd also want MLProgram to be usable along this path.

Where there is no loss of representation or semantics, the more stable form is preferred. Although even there there is a cost to lowering that may not be required for usage.

Could this be a separate pass? E.g., `convert_to_long_term_stable` which incur overheads of conversions to years stable format (which I believe is at least two dialect conversions) and so it is a serialize step.

**Comment by hawkinsp** [comment]
@jpienaar it can be a separate pass, but the question is "what dialect crosses the plugin boundary"? Unless the plugin opts out of stability, that must be a stable dialect. So `chlo` is out at the moment. I'm fine with plugins saying "I don't care about stability" and accepting a superset.

**Comment by burmako** [comment]
"Could this be a separate pass? E.g., convert_to_long_term_stable which incur overheads of conversions to years stable format (which I believe is at least two dialect conversions) and so it is a serialize step". Right, this is the approach that we are taking in [serialize_portable_artifact](https://github.com/tensorflow/tensorflow/blob/a8778f373d19b10f6d270e21426acdc103a76efd/tensorflow/compiler/xla/python/mlir.cc#L161) in `tensorflow/compiler/xla/python` which is used by jax2tf.

**Comment by hawkinsp** [comment]
>  I'm not too familiar with JAX plugins, are they built against different versions of MLIR and we need to manage slightly different versions communicating?

Yes. We expect to use plugins built by third-party vendors using a different version of MLIR.
